src/coprocessor/dag/batch/executors/selection_executor.rs

```diff
@@ -0,0 +1,639 @@
+                .build()],
+        );
+
+        // When source executor returns empty rows, selection executor should process correctly.
+        // No errors should be generated and the predicate function should not be called.
+
+        let r = exec.next_batch(1);
+        // The scan rows parameter has no effect for mock executor. We don't care.
+        assert_eq!(r.data.rows_len(), 0);
+        assert!(!r.is_drained.unwrap());
+
+        let r = exec.next_batch(1);
+        assert_eq!(r.data.rows_len(), 0);
+        assert!(r.is_drained.unwrap());
+    }
+
+    fn make_src_executor_using_fixture_1() -> MockExecutor {
+        MockExecutor::new(
+            vec![FieldTypeTp::LongLong.into(), FieldTypeTp::Double.into()],
+            vec![
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::from(vec![
+                        VectorValue::Int(vec![Some(1), None]),
+                        VectorValue::Real(vec![None, Some(7.0)]),
+                    ]),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(false),
+                },
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::empty(),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(false),
+                },
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::from(vec![
+                        VectorValue::Int(vec![None]),
+                        VectorValue::Real(vec![None]),
+                    ]),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(true),
+                },
+            ],
+        )
+    }
+
+    /// Tests the scenario that there is no predicate or there is a predicate but always returns
+    /// true (no data is filtered).
+    #[test]
+    fn test_no_predicate_or_predicate_always_true() {
+        // Build a selection executor without predicate.
+        let exec_no_predicate =
+            |src_exec: MockExecutor| BatchSelectionExecutor::new_for_test(src_exec, vec![]);
+
+        // Build a selection executor with a predicate that always returns true.
+        let exec_predicate_true = |src_exec: MockExecutor| {
+            let predicate = RpnExpressionBuilder::new()
+                .push_constant(1i64, FieldTypeTp::LongLong)
+                .build();
+            BatchSelectionExecutor::new_for_test(src_exec, vec![predicate])
+        };
+
+        let executor_builders: Vec<Box<dyn std::boxed::FnBox(MockExecutor) -> _>> =
+            vec![Box::new(exec_no_predicate), Box::new(exec_predicate_true)];
+
+        for exec_builder in executor_builders {
+            let src_exec = make_src_executor_using_fixture_1();
+
+            let mut exec = exec_builder(src_exec);
+
+            // The selection executor should return data as it is.
+
+            let r = exec.next_batch(1);
```

Comment:
Maybe compare with which `src_exec.next_batch` returns will be more clear?

Reply:
I wanted to do this, but the result is not cloneable, which makes it impossible to compare.